src/Components/Components/src/RenderTree/StackObjectPool.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    // This is a very simple object pool that requires Get and Return calls to be
+    // balanced as in a stack. It retains up to 'maxPreservedItems' instances in
+    // memory, then for any further requests it supplies untracked instances.
+
+    internal class StackObjectPool<T> where T : class, new()
```

Comment:
I did consider bringing in `ObjectPool` from the extensions repo. However that involves referencing a whole extra assembly and its functionality is much more general than we need. And not that it will matter a lot, but `StackObjectPool` will perform slightly better.

If at some later date we end up using `ObjectPool` in client-side Blazor for other reasons, we can consider replacing `StackObjectPool` with it.

Reply:
Yeah, I really wish that I had made object pool a source package instead of a public api 🥀